apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpanAtomicReference.java

```diff
@@ -38,8 +38,8 @@ public V incrementReferencesAndGet() {
             if (ref != null) {
                 ref.incrementReferences();
             }
+            return ref;
```

Comment:
[For Reviewer]
The return needs to be inside the synchronized, otherwise `ref` might have already been updated by someone else.

This was discovered thanks to the `SpanCompressionIT`.

Reply:
We may need a concurrent load test for this class, the intention for span compression was to avoid locks so that highly concurrent applications don't get penalized and I worry that this might become a bottleneck in that scenario

Reply:
I spent 2 days before the break trying to setup a load test specifically for this, it's difficult. I'm going to review without load testing, and keep a background task for load testing